### PR TITLE
OF-2219: Rewrite of MUC clustering

### PR DESCRIPTION
**Pull request by evdherberg**
There have been a number of MUC clustering problems resolved in Openfire 4.7 back through to 4.2 and possibly before. It's becoming clearer that rather than attempt to patch each issue in turn, that a revised approach on clustering for MUCs is required in order to achieve stability, involving rework to caches, tasks and state.

For background and details please refer to [OF-2219](https://igniterealtime.atlassian.net/browse/OF-2219).

This PR is related to (depends on) changes in Openfire Core. See [that PR](https://github.com/igniterealtime/Openfire/pull/1878) for details.

This PR contains the following features.

### Migrate generic clustering functionality from Hazelcast plugin to Openfire Core

The Hazelcast plugin contained a substantial amount of cache management code that is not specific to Hazelcast clustered caches. This code has been moved out of the plugin, into Openfire Core. This functionality found a new home mainly in `RoutingTableImpl` and `SessionManagerImpl` classes. This should fix https://github.com/igniterealtime/openfire-hazelcast-plugin/issues/50

### CODE REVIEW

**Comment by GregDThomas** [comment]
I've yet to look at this properly (it's sizeable) but can I strongly recommend that before this merge, a new version of the existing plugin is released.

Once merged, the new plugin will only work with the version of Openfire that has https://github.com/igniterealtime/Openfire/pull/1878 in it.

And, the existing version of the plugin will stop working once https://github.com/igniterealtime/Openfire/pull/1878 is merged.

So, for that reason, I suggest another PR that just updates the patch version of the plugin, and sets  `<maxServerVersion>` to 4.6.

This means;
a) users who have yet to upgrade to 4.7 won't accidentally pick up this new version of the plugin that won't work, and
b) users who have upgraded to 4.7 and have the old version of the plugin still installed won't get in a mess (the plugin, and hence clustering, will be disabled until the plugin is updated)

File: src/java/org/jivesoftware/openfire/plugin/util/cache/ClusterListener.java [hunk 1]
```diff
@@ -680,14 +142,16 @@ private synchronized void leaveCluster() {
...
+
         final boolean wasSenior = isSenior;
         isSenior = isSeniorClusterMember();
         // local member only
         final NodeID nodeID = ClusteredCacheFactory.getNodeID(event.getMember());
         if (event.getMember().localMember()) { // We left and re-joined the cluster
             joinCluster();
         } else {
-            nodePresences.put(nodeID, new ConcurrentHashMap<>());
+            ///nodePresences.put(nodeID, new ConcurrentHashMap<>());
```
Comment (GregDThomas) [inline src/java/org/jivesoftware/openfire/plugin/util/cache/ClusterListener.java:154]: Can we simply remove this line if it's not used?

File: src/java/org/jivesoftware/openfire/plugin/util/cache/ClusterListener.java [hunk 2]
```diff
@@ -717,14 +183,13 @@ public void memberRemoved(final MembershipEvent event) {
 
             // Clean up directed presences sent from entities hosted in the leaving node to local entities
             // Clean up directed presences sent to entities hosted in the leaving node from local entities
-            cleanupDirectedPresences(nodeID);
+//            cleanupDirectedPresences(nodeID);
```
Comment (GregDThomas) [inline src/java/org/jivesoftware/openfire/plugin/util/cache/ClusterListener.java:186]: Can we simply remove this line if it's not used?

**Comment by GregDThomas** [comment]
Broadly, looks fine. A few picky comments on commented out code.

But ... 18 commits? Can we squash some of those, please? Not necessarily just one, but 18 seems a bit much to me? 

(But if they won't readily squash, for whatever reason that's not a hard no!).

**Comment by guusdk** [comment]
> I've yet to look at this properly (it's sizeable) but can I strongly recommend that before this merge, a new version of the existing plugin is released.

Daryl did this yesterday. This PR should be rebased.

**Comment by guusdk** [comment]
I'm not bothered much by having plenty of commits - especially if they're illustrating the refactoring process that took place. Maybe we can squash some of the meaningless ones, but otherwise... meh.

**Pull request merged**